AX: Stop calling updateBackingStore() in AXIsolatedObject::children() for performance reasons

twilco opened the pull request:
#### 56d65459b823771b022695053f05ce1bfb7b286d
<pre>
AX: Stop calling updateBackingStore() in AXIsolatedObject::children() for performance reasons
<a href="https://bugs.webkit.org/show_bug.cgi?id=287862">https://bugs.webkit.org/show_bug.cgi?id=287862</a>
<a href="https://rdar.apple.com/145062082">rdar://145062082</a>

Reviewed by Chris Fleizach.

AXIsolatedObject::updateBackingStore() calls AXIsolatedTree::applyPendingChanges(), which takes a lock. This is not
ideal to use in a function that is as hot as AXIsolatedObject::children(). Calling updateBackingStore() also requires
a strong-ref and strong-deref to |this| in the function, since updateBackingStore() can delete |this|. These refs
and derefs, and updateBackingStore show up in samples taken on a popular video website.

With this commit, we remove the call to updateBackingStore in AXIsolatedObject::children. We really should only need
to call updateBackingStore once for every attribute request, and should do so right at the beginning of the attribute
request in the wrapper. This may also be a bit more logical, ensuring each attribute request is served from the same
snapshot of the tree.

This patch also features several unrelated performance improvements:
  - Remove updateChildrenIfNecessary from AXIsolatedObject. It&apos;s a
    no-op, but we still incur the cost of a virtual function call in our
    hottest codepaths (e.g. nextInPreOrder).
  - Remove unnecessary ref-churn, e.g. creating a Ref just to call Vector::find.
  - Inline isValid() and isInTextRun() in AXTextMarker::findMarker, our
    hottest text marker function.
  - Reduce unnecessary calls to AXTextMarker::runs().

All of these things showed up significantly in samples.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::previousInPreOrder):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
(WebCore::AXCoreObject::previousSiblingIncludingIgnored):
(WebCore::AXCoreObject::nextUnignoredSibling const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::appendAccessibilityObject):
(WebCore::appendChildrenToArray):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findMarker const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addNodeOnlyChildren):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::removeChildScrollbar):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::children):
(WebCore::AXIsolatedObject::updateChildrenIfNecessary): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _accessibilityHitTest:returnPlatformElements:]):

Canonical link: <a href="https://commits.webkit.org/290904@main">https://commits.webkit.org/290904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22fabb0941056627027145f818da174ba061ad07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91165 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10708 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96166 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41921 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93215 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11112 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19027 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70057 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27581 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94166 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8470 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50383 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8241 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/171 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41058 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78561 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/186 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98149 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18366 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13467 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79066 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18624 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78428 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78269 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19397 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22782 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/126 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11553 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18367 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23681 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18090 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21551 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19866 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -897,30 +897,35 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, CoalesceObjectBreak
-    if (!isInTextRun())
+    }
+    const auto* runs = object->textRuns();
+    if (!runs) {
```
Does this also need to check `runs->size()` like `isInTextRun()` does?

twilco replied:
Oops, yes!

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -177,7 +177,9 @@ void AccessibilityScrollView::removeChildScrollbar(AccessibilityObject* scrollba
         return;
 
-    size_t position = m_children.find(Ref { *scrollbar });
+    size_t position = m_children.findIf([&] (const Ref<AXCoreObject>& child) {
```
nit: We might want to be specific here with this lambda and pass in the reference to scrollbar specifically:
```c++
size_t position = m_children.findIf([&scrollbar] (const Ref<AXCoreObject>& child) {
```

twilco committed:
AX: Stop calling updateBackingStore() in AXIsolatedObject::children() for performance reasons
https://bugs.webkit.org/show_bug.cgi?id=287862
rdar://145062082

Reviewed by Chris Fleizach.

AXIsolatedObject::updateBackingStore() calls AXIsolatedTree::applyPendingChanges(), which takes a lock. This is not
ideal to use in a function that is as hot as AXIsolatedObject::children(). Calling updateBackingStore() also requires
a strong-ref and strong-deref to |this| in the function, since updateBackingStore() can delete |this|. These refs
and derefs, and updateBackingStore show up in samples taken on a popular video website.

With this commit, we remove the call to updateBackingStore in AXIsolatedObject::children. We really should only need
to call updateBackingStore once for every attribute request, and should do so right at the beginning of the attribute
request in the wrapper. This may also be a bit more logical, ensuring each attribute request is served from the same
snapshot of the tree.

This patch also features several unrelated performance improvements:
  - Remove updateChildrenIfNecessary from AXIsolatedObject. It's a
    no-op, but we still incur the cost of a virtual function call in our
    hottest codepaths (e.g. nextInPreOrder).
  - Remove unnecessary ref-churn, e.g. creating a Ref just to call Vector::find.
  - Inline isValid() and isInTextRun() in AXTextMarker::findMarker, our
    hottest text marker function.
  - Reduce unnecessary calls to AXTextMarker::runs().

All of these things showed up significantly in samples.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::previousInPreOrder):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
(WebCore::AXCoreObject::previousSiblingIncludingIgnored):
(WebCore::AXCoreObject::nextUnignoredSibling const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::appendAccessibilityObject):
(WebCore::appendChildrenToArray):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findMarker const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addNodeOnlyChildren):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::removeChildScrollbar):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::children):
(WebCore::AXIsolatedObject::updateChildrenIfNecessary): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _accessibilityHitTest:returnPlatformElements:]):

Canonical link: https://commits.webkit.org/290904@main
